### PR TITLE
Miner: Efficiently expire pre-committed sectors

### LINKED ISSUE #417: Quantize sector pre-commit expiry epochs to avoid per-sector cron registration
When each sector is pre-committed, a cron event is registered for the expiry of that precommit, which penalises a miner for failing to follow through. This creates a huge load of cron registrations in state, and processing in the tipset cron handler. For high-scale miners, many events per epoch. Most of the events are for naught, but can't be unregistered when the proof arrives on time.

This load could reduced by quantizing the epochs at which expiries are processed, and using a single cron event per miner per quanta which checks all the pre-commits due for proof before the event. The existing proving-period event might be a suitable entry point. It doesn't matter much if it's quite infrequent: a late-arriving proof is already rejected. This is just cleaning up state.

The precommit sector IDs for processing could be stored either in a queue-like `AMT[epoch]Bitfield`, perhaps quantized (like `FaultEpochs`)

### PR DESCRIPTION
**Pull request by aarshkshah1992**
Closes #417.

### CODE REVIEW

File: actors/builtin/miner/miner_actor.go [hunk 1]
```diff
@@ -498,25 +497,31 @@ func (a Actor) PreCommitSector(rt Runtime, params *SectorPreCommitInfo) *adt.Emp
...
 	// ConfirmSectorProofsValid would fail to find it.
 	expiryBound := rt.CurrEpoch() + msd + 1
-	enrollCronEvent(rt, expiryBound, &cronPayload)
+
+	// create BitField for this sector
+	bf := abi.NewBitField()
+	bf.Set(uint64(params.SectorNumber))
+
+	// Load BitField Queue for sector expiry
+	quant := st.QuantEndOfDeadline()
```
Comment (Stebalien) [inline actors/builtin/miner/miner_actor.go:515]: This has gone away. We'll need to quantize for every deadline, at the end of every deadline.

File: actors/builtin/miner/miner_actor.go [hunk 2]
```diff
@@ -498,25 +497,31 @@ func (a Actor) PreCommitSector(rt Runtime, params *SectorPreCommitInfo) *adt.Emp
...
+	queue, err := LoadBitfieldQueue(store, st.PreCommittedSectorsExpiry, quant)
+	builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to load pre-commit expiry queue")
+
+	// add entry for this sector to the queue
+	err = queue.AddToQueue(expiryBound, bf)
```
Comment (Stebalien) [inline actors/builtin/miner/miner_actor.go:520]: nit:

```suggestion
	err = queue.AddToQueueValues(expiryBound, uint64(params.SectorNumber))
```

File: actors/builtin/miner/miner_actor.go [hunk 3]
```diff
@@ -1610,6 +1611,24 @@ func handleProvingDeadline(rt Runtime) {
...
+
+		bf, modified, err := expiryQ.PopUntil(currEpoch)
+		builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to pop expired sectors")
+
+		if modified {
+			st.PreCommittedSectorsExpiry, err = expiryQ.Root()
+			builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to save expiry queue")
+		}
+
+		if bf != nil {
```
Comment (Stebalien) [inline actors/builtin/miner/miner_actor.go:1627]: this should never return nil, unless it returns an error.

File: actors/builtin/miner/miner_actor.go [hunk 4]
```diff
@@ -1610,6 +1611,24 @@ func handleProvingDeadline(rt Runtime) {
...
+			depositToBurn, err := st.checkPrecommitExpiry(store, bf)
+			builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to expire pre-committed sectors")
+			burnFunds(rt, depositToBurn)
```
Comment (Stebalien) [inline actors/builtin/miner/miner_actor.go:1630]: Can't call this in a transaction.

Comment (Stebalien) [reply actors/builtin/miner/miner_actor.go:1630]: This should probably just be added to `penaltyTotal` and burnt at the end.

File: actors/builtin/miner/miner_actor.go [hunk 5]
```diff
@@ -498,25 +497,31 @@ func (a Actor) PreCommitSector(rt Runtime, params *SectorPreCommitInfo) *adt.Emp
 
 	notifyPledgeChanged(rt, newlyVested.Neg())
 
-	bf := abi.NewBitField()
-	bf.Set(uint64(params.SectorNumber))
-
-	// Request deferred Cron check for PreCommit expiry check.
-	cronPayload := CronEventPayload{
-		EventType: CronEventPreCommitExpiry,
-		Sectors:   bf,
```
Comment (Stebalien) [inline actors/builtin/miner/miner_actor.go:507]: We should be able to remove `Sectors` from the cron payload.

File: actors/builtin/miner/miner_actor.go (hunk 1 above)
Comment (anorth) [reply actors/builtin/miner/miner_actor.go:515]: `NewQuantSpec(WPoStChallengeWindow, st.ProvingPeriodStart)`

File: actors/builtin/miner/miner_actor.go [hunk 6]
```diff
@@ -498,25 +497,31 @@ func (a Actor) PreCommitSector(rt Runtime, params *SectorPreCommitInfo) *adt.Emp
...
+	quant := st.QuantEndOfDeadline()
+	queue, err := LoadBitfieldQueue(store, st.PreCommittedSectorsExpiry, quant)
+	builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to load pre-commit expiry queue")
+
+	// add entry for this sector to the queue
+	err = queue.AddToQueue(expiryBound, bf)
+	builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to add pre-commit sector expiry to queue")
+
+	st.PreCommittedSectorsExpiry, err = queue.Root()
+	builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to save pre-commit sector queue")
```
Comment (anorth) [inline actors/builtin/miner/miner_actor.go:524]: Please push this whole delta down to a method on miner_state, and test it directly.

File: actors/builtin/miner/miner_actor.go [hunk 7]
```diff
@@ -1610,6 +1611,24 @@ func handleProvingDeadline(rt Runtime) {
...
+		if modified {
+			st.PreCommittedSectorsExpiry, err = expiryQ.Root()
+			builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to save expiry queue")
+		}
+
+		if bf != nil {
+			depositToBurn, err := st.checkPrecommitExpiry(store, bf)
+			builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to expire pre-committed sectors")
+			burnFunds(rt, depositToBurn)
+		}
```
Comment (anorth) [inline actors/builtin/miner/miner_actor.go:1631]: Let's move this whole lot to the top, immediately after vesting funds.

Push it down into a method on miner.State that returns the total penalty.

File: actors/builtin/miner/miner_actor.go [hunk 8]
```diff
@@ -1686,46 +1705,44 @@ func validateReplaceSector(rt Runtime, st *State, store adt.Store, params *Secto
 	return replaceSector
 }
 
-func checkPrecommitExpiry(rt Runtime, sectors *abi.BitField) {
-	store := adt.AsStore(rt)
-	var st State
-
+func (st *State) checkPrecommitExpiry(store adt.Store, sectors *abi.BitField) (depositToBurn abi.TokenAmount, err error) {
```
Comment (anorth) [inline actors/builtin/miner/miner_actor.go:1708]: Move to miner_state.go

File: actors/builtin/miner/miner_actor.go [hunk 9]
```diff
@@ -1686,46 +1705,44 @@ func validateReplaceSector(rt Runtime, st *State, store adt.Store, params *Secto
...
-				return nil
-			}
+	depositToBurn = abi.NewTokenAmount(0)
 
-			// mark it for deletion
-			sectorNos = append(sectorNos, sectorNo)
-
-			// increment deposit to burn
-			depositToBurn = big.Add(depositToBurn, sector.PreCommitDeposit)
+	var sectorNos []abi.SectorNumber
```
Comment (anorth) [inline actors/builtin/miner/miner_actor.go:1712]:
```suggestion
	var precommitsToDelete []abi.SectorNumber
```

File: actors/builtin/miner/miner_actor.go (hunk 6 above)
Comment (aarshkshah1992) [reply actors/builtin/miner/miner_actor.go]: This is done.

File: actors/builtin/miner/miner_actor.go (hunk 3 above)
Comment (aarshkshah1992) [reply actors/builtin/miner/miner_actor.go]: Have removed this check and added an Assert for this.

File: actors/builtin/miner/miner_actor.go (hunk 7 above)
Comment (aarshkshah1992) [reply actors/builtin/miner/miner_actor.go]: @anorth Not sure what you mean here. Do you mean push the pre commit expiry part in a separate func ?

File: actors/builtin/miner/miner_actor.go (hunk 4 above)
Comment (aarshkshah1992) [reply actors/builtin/miner/miner_actor.go]: This is done.

File: actors/builtin/miner/miner_state.go [hunk 10]
```diff
@@ -970,6 +975,81 @@ func (st *State) MeetsInitialPledgeCondition(balance abi.TokenAmount) bool {
...
+// pre-commit expiry
+func (st *State) AddPreCommitExpiry(store adt.Store, currEpoch abi.ChainEpoch, sealProof abi.RegisteredSealProof, sectorNum abi.SectorNumber) error {
+	msd, ok := MaxSealDuration[sealProof]
+	if !ok {
+		return xerrors.Errorf("no max seal duration set for proof type: %d", sealProof)
+	}
+	// The +1 here is critical for the batch verification of proofs. Without it, if a proof arrived exactly on the
+	// due epoch, ProveCommitSector would accept it, then the expiry event would remove it, and then
+	// ConfirmSectorProofsValid would fail to find it.
+	expiryBound := currEpoch + msd + 1
```
Comment (anorth) [inline actors/builtin/miner/miner_state.go:987]: Nit: pull this lookup and addition back out to the actor, separating the policy (in the actor) from the mechanics here.

File: actors/builtin/miner/miner_state.go [hunk 11]
```diff
@@ -970,6 +975,81 @@ func (st *State) MeetsInitialPledgeCondition(balance abi.TokenAmount) bool {
...
+
+	// create BitField for this sector
+	bf := abi.NewBitField()
+	bf.Set(uint64(sectorNum))
```
Comment (anorth) [inline actors/builtin/miner/miner_state.go:991]: Delete this bf, unused.

File: actors/builtin/miner/miner_state.go [hunk 12]
```diff
@@ -970,6 +975,81 @@ func (st *State) MeetsInitialPledgeCondition(balance abi.TokenAmount) bool {
...
+
+	// Load BitField Queue for sector expiry
+	quant := NewQuantSpec(WPoStChallengeWindow, st.ProvingPeriodStart)
```
Comment (anorth) [inline actors/builtin/miner/miner_state.go:994]: We create this twice, so extract `st.QuantSpecEveryDeadline()`.

File: actors/builtin/miner/miner_state.go [hunk 13]
```diff
@@ -970,6 +975,81 @@ func (st *State) MeetsInitialPledgeCondition(balance abi.TokenAmount) bool {
...
+	}
+
+	// add entry for this sector to the queue
+	if err := queue.AddToQueueValues(expiryBound, uint64(sectorNum)); err != nil {
+		return xerrors.Errorf("failed to add pre-commit sector expiry to queue: %w", err)
+	}
+
+	st.PreCommittedSectorsExpiry, err = queue.Root()
+	if err != nil {
+		return xerrors.Errorf("failed to save pre-commit sector queue")
```
Comment (anorth) [inline actors/builtin/miner/miner_state.go:1007]: Wrap the error

File: actors/builtin/miner/miner_state.go [hunk 14]
```diff
@@ -970,6 +975,81 @@ func (st *State) MeetsInitialPledgeCondition(balance abi.TokenAmount) bool {
...
+	}
+
+	return nil
+}
+
+func (st *State) checkPrecommitExpiry(store adt.Store, sectors *abi.BitField) (depositToBurn abi.TokenAmount, err error) {
+	// initialize here to add together for all sectors and minimize calls across actors
```
Comment (anorth) [inline actors/builtin/miner/miner_state.go:1014]: This comment is out of place now, there are no actor calls

File: actors/builtin/miner/miner_state.go (hunk 10 above)
Comment (aarshkshah1992) [reply actors/builtin/miner/miner_state.go:987]: This is done.

File: actors/builtin/miner/miner_state.go (hunk 11 above)
Comment (aarshkshah1992) [reply actors/builtin/miner/miner_state.go:991]: This is done.

File: actors/builtin/miner/miner_state.go (hunk 13 above)
Comment (aarshkshah1992) [reply actors/builtin/miner/miner_state.go:1007]: This is done.

File: actors/builtin/miner/miner_state.go (hunk 14 above)
Comment (aarshkshah1992) [reply actors/builtin/miner/miner_state.go:1014]: Removed.

File: actors/builtin/miner/miner_state.go (hunk 12 above)
Comment (aarshkshah1992) [reply actors/builtin/miner/miner_state.go:994]: This is done.

**Pull request merged**